From 723e9d997ec9dbf6517b64414cffc127a36ad25b Mon Sep 17 00:00:00 2001
From: Ben Kochie <superq@gmail.com>
Date: Tue, 2 Mar 2021 16:46:47 +0100
Subject: [PATCH] Sanitize strings from /sys/class/power_supply

Avoid panic on invalid UTF-8 from /sys/class/power_supply by
sanitizing strings parsed from the kernel.
* Add a broken string to the test fixtures.

Fixes: https://github.com/prometheus/node_exporter/issues/1979

Signed-off-by: Ben Kochie <superq@gmail.com>
---
 CHANGELOG.md                  | 2 ++
 collector/fixtures/sys.ttar   | 2 +-
 collector/powersupplyclass.go | 3 ++-
 3 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/CHANGELOG.md b/CHANGELOG.md
index fff87d731..e8def3f73 100644
--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,8 @@
 * [ENHANCEMENT]
 * [BUGFIX]
 
+* [BUGFIX] Sanitize strings from /sys/class/power_supply
+
 ## 1.1.1 / 2021-02-12
 
 * [BUGFIX] Fix ineffassign issue #1957
diff --git a/collector/fixtures/sys.ttar b/collector/fixtures/sys.ttar
index 0392d2050..9f0724309 100644
--- a/collector/fixtures/sys.ttar
+++ b/collector/fixtures/sys.ttar
@@ -1298,7 +1298,7 @@ Mode: 444
 # ttar - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 Path: sys/class/power_supply/BAT0/model_name
 Lines: 1
-LNV-45N1
+LNV-45N1¿¿
 Mode: 444
 # ttar - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
 Directory: sys/class/power_supply/BAT0/power
diff --git a/collector/powersupplyclass.go b/collector/powersupplyclass.go
index adebf0e6b..f20dab4d4 100644
--- a/collector/powersupplyclass.go
+++ b/collector/powersupplyclass.go
@@ -21,6 +21,7 @@ import (
 	"fmt"
 	"os"
 	"regexp"
+	"strings"
 
 	"github.com/go-kit/kit/log"
 	"github.com/prometheus/client_golang/prometheus"
@@ -153,7 +154,7 @@ func (c *powerSupplyClassCollector) Update(ch chan<- prometheus.Metric) error {
 		} {
 			if value != "" {
 				keys = append(keys, name)
-				values = append(values, value)
+				values = append(values, strings.ToValidUTF8(value, ""))
 			}
 		}
 
